File: apps/driver-android/src/main/kotlin/lk/mageride/driver/vehicle/VehicleOnboardingScreen.kt
package lk.mageride.driver.vehicle

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.ArrowBack
import androidx.compose.material.icons.outlined.AutoAwesome
import androidx.compose.material.icons.outlined.Info
import androidx.compose.material.icons.outlined.WarningAmber
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ExposedDropdownMenuAnchorType
import androidx.compose.material3.ExposedDropdownMenuBox
import androidx.compose.material3.ExposedDropdownMenuDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import lk.mageride.driver.R
import lk.mageride.driver.capture.DocumentCaptureTarget
import lk.mageride.driver.ui.component.AdminVerifyChip
import lk.mageride.driver.ui.component.CaptureTile
import lk.mageride.driver.ui.component.ExtractedFieldRow
import lk.mageride.driver.ui.component.MageRideCta
import lk.mageride.driver.ui.component.ModeCBadge
import lk.mageride.driver.ui.component.NoticeCard
import lk.mageride.driver.ui.component.StatusPill
import lk.mageride.driver.ui.component.StatusTone
import lk.mageride.driver.ui.labelRes
import lk.mageride.driver.ui.theme.ControlTokens
import lk.mageride.driver.ui.theme.MageRideTheme
import lk.mageride.shared.data.models.ExtractedField
import lk.mageride.shared.data.models.RideVehicleType
import lk.mageride.shared.data.models.registry.OnboardingStep
import org.koin.androidx.compose.koinViewModel
import kotlin.math.roundToInt

/**
 * **SCR-DA-004 → 004c · vehicle onboarding** — the optional Mode-C four-step wizard.
 *
 * One route and one composable for all four steps, because that is what they are: the wireframe
 * draws the same app bar, the same progress bar and the same CTA on each, with a different body
 * between them. Four destinations would put the resume rule (AL-30) in the navigation graph
 * instead of in [VehicleOnboardingViewModel], where it can be tested.
 *
 * **There is no permit field and no GPS-tracker field on any of the four.** Mode A/B vehicles and
 * route permits are onboarded in the Fleet Portal (AL-27, SCR-FP-004) and the wizard says so, in
 * the notice card the wireframe puts on Step 1/4.
 *
 * @param onCaptureRequested Opens SCR-DA-005 — the tiles here never photograph anything.
 * @param onSubmitted Step 4/4 is saved; hand over to SCR-DA-006.
 * @param onExit Back from Step 1/4 leaves the wizard entirely.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun VehicleOnboardingScreen(
    onCaptureRequested: () -> Unit,
    onSubmitted: () -> Unit,
    onExit: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val viewModel: VehicleOnboardingViewModel = koinViewModel()
    val state by viewModel.state.collectAsStateWithLifecycle()

    LaunchedEffect(state.submitted) {
        if (state.submitted) onSubmitted()
    }
    LaunchedEffect(state.exited) {
        if (state.exited) onExit()
    }

    Scaffold(
        modifier = modifier.fillMaxSize(),
        topBar = {
            TopAppBar(
                title = { Text(text = stringResource(state.step.titleRes())) },
                navigationIcon = {
                    IconButton(onClick = viewModel::onBack) {
                        Icon(
                            imageVector = Icons.AutoMirrored.Outlined.ArrowBack,
                            contentDescription = stringResource(R.string.action_back),
                        )
                    }
                },
                actions = {
                    ModeCBadge(
                        label = stringResource(R.string.mode_c_badge),
                        modifier = Modifier.padding(end = MageRideTheme.spacing.sm),
                    )
                },
            )
        },
    ) { insets ->
        if (state.loading) {
            Box(
                modifier = Modifier
                    .padding(insets)
                    .fillMaxSize(),
                contentAlignment = Alignment.Center,
            ) {
                CircularProgressIndicator()
            }
            return@Scaffold
        }

        Column(
            modifier = Modifier
                .padding(insets)
                .fillMaxSize()
                .verticalScroll(rememberScrollState())
                .padding(MageRideTheme.spacing.md),
            verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.sm),
        ) {
            LinearProgressIndicator(progress = { state.progress }, modifier = Modifier.fillMaxWidth())
            Text(
                text = stringResource(
                    R.string.vehicle_onboard_step_of,
                    state.stepNumber,
                    OnboardingStep.entries.size,
                    stringResource(state.step.captionRes()),
                ),
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )

            when (state.step) {
                OnboardingStep.DETAILS -> DetailsStep(state = state, viewModel = viewModel)

                OnboardingStep.INSURANCE -> DocumentStep(
                    state = state,
                    target = DocumentCaptureTarget.INSURANCE,
                    slotLabel = stringResource(R.string.vehicle_onboard_insurance_slot),
                    onCapture = {
                        viewModel.requestCapture(DocumentCaptureTarget.INSURANCE)
                        onCaptureRequested()
                    },
                )

                OnboardingStep.REVENUE -> DocumentStep(
                    state = state,
                    target = DocumentCaptureTarget.REVENUE_LICENCE,
                    slotLabel = stringResource(R.string.vehicle_onboard_revenue_slot),
                    onCapture = {
                        viewModel.requestCapture(DocumentCaptureTarget.REVENUE_LICENCE)
                        onCaptureRequested()
                    },
                )

                OnboardingStep.PHOTOS -> PhotosStep(
                    state = state,
                    onCapture = { target ->
                        viewModel.requestCapture(target)
                        onCaptureRequested()
                    },
                )
            }

            if (state.isPendingReview) {
                PendingReviewCard()
            }

            state.error?.let { message ->
                Text(
                    text = stringResource(message),
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.error,
                )
            }

            MageRideCta(
                label = stringResource(state.step.ctaRes()),
                enabled = state.canContinue,
                loading = state.busy,
                onClick = viewModel::onContinue,
                modifier = Modifier.padding(top = MageRideTheme.spacing.xs),
            )
        }
    }
}

/**
 * **SCR-DA-004 · Step 1/4** — vehicle type and registration number, and nothing else.
 *
 * The notice card is not decoration. It is the one place a Mode-C driver is told where a permit
 * and a Mode A/B vehicle go, and the absence of those two fields is the AL-27 fence itself.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun DetailsStep(
    state: VehicleOnboardingState,
    viewModel: VehicleOnboardingViewModel,
    modifier: Modifier = Modifier,
) {
    var expanded by remember { mutableStateOf(false) }

    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.sm),
    ) {
        ExposedDropdownMenuBox(expanded = expanded, onExpandedChange = { expanded = it }) {
            OutlinedTextField(
                value = state.vehicleType?.let { stringResource(it.labelRes()) }.orEmpty(),
                onValueChange = {},
                modifier = Modifier
                    .fillMaxWidth()
                    // `PrimaryNotEditable` — the field is the menu's anchor and is read-only, so
                    // TalkBack announces a dropdown rather than a text box that will not accept
                    // typing. The no-argument `menuAnchor()` is deprecated in M3 1.4.
                    .menuAnchor(ExposedDropdownMenuAnchorType.PrimaryNotEditable),
                readOnly = true,
                label = { Text(text = stringResource(R.string.vehicle_onboard_type_label)) },
                trailingIcon = { ExposedDropdownMenuDefaults.TrailingIcon(expanded = expanded) },
            )
            ExposedDropdownMenu(expanded = expanded, onDismissRequest = { expanded = false }) {
                // The Mode-C subset (AL-09). `bus` and `train` are Mode A and are not offered —
                // `POST /v1/vehicles` answers 403 mode-not-allowed for either.
                RideVehicleType.entries.forEach { type ->
                    DropdownMenuItem(
                        text = { Text(text = stringResource(type.labelRes())) },
                        onClick = {
                            viewModel.onVehicleTypeChanged(type)
                            expanded = false
                        },
                    )
                }
            }
        }

        OutlinedTextField(
            value = state.registrationNumber,
            onValueChange = viewModel::onRegistrationChanged,
            modifier = Modifier.fillMaxWidth(),
            label = { Text(text = stringResource(R.string.vehicle_onboard_registration_label)) },
            singleLine = true,
            isError = state.registrationTaken,
            supportingText = {
                if (state.registrationTaken) {
                    Text(
                        text = stringResource(R.string.error_registration_exists),
                        style = MaterialTheme.typography.labelSmall,
                    )
                }
            },
        )

        NoticeCard(
            accent = MageRideTheme.mode.modeC,
            icon = Icons.Outlined.Info,
            title = stringResource(R.string.vehicle_onboard_mode_c_title),
        ) {
            Text(
                text = stringResource(R.string.vehicle_onboard_mode_c_body),
                style = MaterialTheme.typography.bodyMedium,
            )
        }
    }
}

/**
 * **SCR-DA-004a / 004b · Steps 2 and 3** — one capture, one `Done ✓`, one extract card.
 *
 * The two steps differ only in what they capture and which fields come back, so they are one
 * composable. The extract card appears only after a save, because the save is what queues the
 * Gemini Flash extraction — see [VehicleOnboardingViewModel].
 */
@Composable
private fun DocumentStep(
    state: VehicleOnboardingState,
    target: DocumentCaptureTarget,
    slotLabel: String,
    onCapture: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.sm),
    ) {
        CaptureTile(
            label = slotLabel,
            captureHint = stringResource(R.string.action_tap_to_capture),
            doneLabel = stringResource(R.string.status_done),
            captured = state.captured(target),
            onClick = onCapture,
            height = ControlTokens.CapturePanel,
        )

        SlotStatusRow(label = slotLabel, done = state.captured(target))
        ExtractionCard(fields = state.stepFields)
    }
}

/**
 * **SCR-DA-004c · Step 4/4** — front and back, number plate visible on both.
 *
 * The notice under the tiles is the step's whole rule: the plate read out of these photos is
 * matched against the registration number typed on Step 1/4, and a mismatch is what makes the step
 * Pending (BR-25.3, US-2.27).
 */
@Composable
private fun PhotosStep(
    state: VehicleOnboardingState,
    onCapture: (DocumentCaptureTarget) -> Unit,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.sm),
    ) {
        Row(horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs)) {
            CaptureTile(
                label = stringResource(R.string.vehicle_onboard_photo_front),
                captureHint = stringResource(R.string.action_tap_to_capture),
                doneLabel = stringResource(R.string.status_done),
                captured = state.photoFront != null,
                onClick = { onCapture(DocumentCaptureTarget.VEHICLE_FRONT) },
                modifier = Modifier.weight(1f),
                height = ControlTokens.CapturePanel,
            )
            CaptureTile(
                label = stringResource(R.string.vehicle_onboard_photo_back),
                captureHint = stringResource(R.string.action_tap_to_capture),
                doneLabel = stringResource(R.string.status_done),
                captured = state.photoBack != null,
                onClick = { onCapture(DocumentCaptureTarget.VEHICLE_BACK) },
                modifier = Modifier.weight(1f),
                height = ControlTokens.CapturePanel,
            )
        }

        SlotStatusRow(
            label = stringResource(R.string.vehicle_onboard_photos_slot),
            done = state.photoFront != null && state.photoBack != null,
        )

        NoticeCard(accent = MageRideTheme.mode.modeC, icon = Icons.Outlined.Info) {
            Text(
                text = stringResource(R.string.vehicle_onboard_plate_match, state.registrationNumber),
                style = MaterialTheme.typography.bodyMedium,
            )
        }

        ExtractionCard(fields = state.stepFields)
    }
}

/** The wireframe's `🛡 Insurance document — Done ✓` row under a capture panel. */
@Composable
private fun SlotStatusRow(label: String, done: Boolean, modifier: Modifier = Modifier) {
    Row(
        modifier = modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(
            text = label,
            modifier = Modifier.weight(1f),
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurface,
        )
        StatusPill(
            label = stringResource(if (done) R.string.status_done else R.string.status_not_captured),
            tone = if (done) StatusTone.DONE else StatusTone.NEUTRAL,
        )
    }
}

/**
 * The wireframe's *"✦ AI-extracted (Gemini Flash 3.0)"* card.
 *
 * **Every row is read-only, and the wireframe draws a ✎ on some of them.** The multipart arm of
 * `PUT /v1/vehicles/{id}/onboarding/{step}` carries `registrationNumber`, `vehicleType`, `file`
 * and `fileBack` — it has **no `fields` part** — and the JSON arm that does needs a `fileId` no
 * response on this surface ever returns. So a correction typed here would have nowhere to go and
 * would be silently dropped on save. Showing the verdicts with the ⚑ chip is the honest
 * rendering, and it is not a dead end: a doubtful field is `pending_review` and goes to the
 * Verification Officer, which is exactly what BR-25.3 asks for. Raised as a micro-change-set in
 * the C069 handoff; same shape as C068's finding on `licence_no` / `licence_expiry`.
 */
@Composable
private fun ExtractionCard(fields: List<ExtractedField>, modifier: Modifier = Modifier) {
    if (fields.isEmpty()) return

    NoticeCard(
        accent = MageRideTheme.status.success,
        icon = Icons.Outlined.AutoAwesome,
        title = stringResource(R.string.vehicle_onboard_extract_title),
        modifier = modifier,
    ) {
        fields.forEach { field ->
            ExtractedFieldRow(
                label = stringResource(field.labelRes()),
                value = field.displayValue(),
                emptyLabel = stringResource(R.string.profile_setup_field_unread),
                flagLabel = stringResource(R.string.flag_admin_verify).takeIf { field.needsOfficerReview },
            )
        }

        // The wireframe's `Confidence · 0.62 — doubtful` row. The **lowest** of the step's fields,
        // because a step is only as verified as its weakest one (BR-25.3), and "doubtful" is read
        // off the server's own `verifyStatus` rather than compared against a threshold here —
        // `Registry:OcrConfidenceThreshold` is admin-tunable and a second copy of it in the app
        // would eventually disagree with the verdict printed beside it.
        fields.lowestConfidence()?.let { confidence ->
            ExtractedFieldRow(
                label = stringResource(R.string.vehicle_field_confidence),
                value = if (fields.any { it.needsOfficerReview }) {
                    stringResource(R.string.vehicle_confidence_doubtful, confidence.formatted())
                } else {
                    confidence.formatted()
                },
                emptyLabel = stringResource(R.string.profile_setup_field_unread),
            )
        }
    }
}

/** The weakest extraction in this step, or `null` when nothing came back with a confidence. */
private fun List<ExtractedField>.lowestConfidence(): Double? = mapNotNull(ExtractedField::confidence).minOrNull()

/** `0.62`. Two decimals, built by arithmetic so the decimal separator is the same in all three locales. */
private fun Double.formatted(): String {
    val hundredths = (this * PERCENT).roundToInt()
    return "${hundredths / PERCENT}.${(hundredths % PERCENT).toString().padStart(2, '0')}"
}

private const val PERCENT = 100

/** The wireframe's amber *"Any element doubtful or edited → this step's status is Pending"* card. */
@Composable
private fun PendingReviewCard(modifier: Modifier = Modifier) {
    NoticeCard(
        accent = MageRideTheme.status.warning,
        icon = Icons.Outlined.WarningAmber,
        title = stringResource(R.string.vehicle_onboard_pending_title),
        modifier = modifier,
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text(
                text = stringResource(R.string.vehicle_onboard_pending_body),
                modifier = Modifier.weight(1f),
                style = MaterialTheme.typography.bodyMedium,
            )
            AdminVerifyChip(label = stringResource(R.string.flag_admin_verify))
        }
    }
}

File: apps/driver-ios/DriverApp/Vehicle/VehicleOnboardingScreen.swift
import MageRideShared
import SwiftUI

/// **SCR-DI-004 → 004c · vehicle onboarding** — the optional Mode-C four-step wizard.
///
/// One route and one view for all four steps, because that is what they are: the wireframe draws the
/// same navigation bar, the same progress bar and the same CTA on each, with a different body between
/// them. Four destinations would put the resume rule (AL-30) in the navigation graph instead of in
/// ``VehicleOnboardingModel``, where it can be tested.
///
/// **There is no permit field and no GPS-tracker field on any of the four.** Mode A/B vehicles and
/// route permits are onboarded in the Fleet Portal (AL-27, SCR-FP-004) and the wizard says so, in the
/// notice card the wireframe puts on Step 1/4.
///
/// - Parameters:
///   - onCaptureRequested: Opens SCR-DI-005 — the tiles here never photograph anything.
///   - onSubmitted: Step 4/4 is saved; hand over to SCR-DI-006.
///   - onExit: Back from Step 1/4 leaves the wizard entirely.
///
/// `@MainActor` on the whole view, not on its initialiser — see ``ProfileSetupScreen`` for why.
@MainActor
struct VehicleOnboardingScreen: View {

    @StateObject private var model: VehicleOnboardingModel
    @ObservedObject private var captures: DocumentCaptureCoordinator

    private let onCaptureRequested: () -> Void
    private let onSubmitted: () -> Void
    private let onExit: () -> Void

    init(
        vehicles: VehicleOnboardingRepository,
        captures: DocumentCaptureCoordinator,
        session: VehicleOnboardingSession,
        onCaptureRequested: @escaping () -> Void,
        onSubmitted: @escaping () -> Void,
        onExit: @escaping () -> Void
    ) {
        _model = StateObject(
            wrappedValue: VehicleOnboardingModel(vehicles: vehicles, captures: captures, session: session)
        )
        self.captures = captures
        self.onCaptureRequested = onCaptureRequested
        self.onSubmitted = onSubmitted
        self.onExit = onExit
    }

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: MageRideSpacing.sm) {
                if model.state.isLoading {
                    ProgressView()
                        .frame(maxWidth: .infinity)
                        .padding(.vertical, MageRideSpacing.xl)
                } else {
                    StepProgress(
                        step: model.state.stepNumber,
                        count: VehicleOnboardingSteps.count,
                        captionKey: model.state.step.captionKey
                    )

                    stepBody

                    if model.state.isPendingReview {
                        pendingReviewCard
                    }
                }

                if let errorKey = model.state.errorKey {
                    FormErrorText(messageKey: errorKey)
                }

                Button(action: { Task { await model.onContinue() } }) {
                    Text(key: model.state.step.ctaKey)
                }
                .buttonStyle(.mageCta(loading: model.state.isBusy))
                .disabled(!model.state.canContinue)
                .padding(.top, MageRideSpacing.xs)
            }
            .padding(MageRideSpacing.md)
        }
        .background(MageRideColor.surface)
        .navigationTitle(Text(key: model.state.step.titleKey))
        .navigationBarTitleDisplayMode(.inline)
        // The wizard's Back is the model's — from Step 1/4 it leaves, and from anywhere else it is a
        // step backwards (D2' §SCR-DI-004). The system's own back button would pop the whole screen
        // from Step 3.
        .navigationBarBackButtonHidden(true)
        .toolbar {
            ToolbarItem(placement: .navigationBarLeading) { backButton }
            ToolbarItem(placement: .navigationBarTrailing) { stepBadge }
        }
        .task { await model.load() }
        // SCR-DI-005 hands the de-skewed image back through the coordinator. Observed rather than
        // passed as a navigation result, because the route carries no arguments.
        .onChange(of: captures.result) { result in
            if let result { model.apply(result) }
        }
        .onChange(of: model.state.isSubmitted) { isSubmitted in
            if isSubmitted { onSubmitted() }
        }
        .onChange(of: model.state.hasExited) { hasExited in
            if hasExited { onExit() }
        }
    }

    // MARK: - The bar

    /// The wireframe's `‹ Cancel` on Step 1/4 and `‹ Back` on the other three — which is the same
    /// distinction the model draws, spelled where the driver can read it before they tap.
    private var backButton: some View {
        Button(action: model.onBack) {
            HStack(spacing: 2) {
                Image(systemName: "chevron.left")
                Text(key: model.state.step == OnboardingStep.details ? "action_cancel" : "action_back")
                    .mageFont(.body)
            }
        }
    }

    /// `Mode C` on Step 1/4, and the wireframe's `2/4` counter afterwards.
    ///
    /// The counter is composed rather than translated: it is two numbers and a solidus, which is data
    /// in every language — the same rule that keeps `+94` and the language endonyms out of the
    /// strings files. The trilingual form of the same fact is the caption under the progress bar.
    @ViewBuilder
    private var stepBadge: some View {
        if model.state.step == OnboardingStep.details {
            ModeCBadge()
        } else {
            Text("\(model.state.stepNumber)/\(VehicleOnboardingSteps.count)")
                .mageFont(.label)
                .foregroundStyle(MageRideColor.onSurfaceVariant)
        }
    }

    // MARK: - The four bodies

    @ViewBuilder
    private var stepBody: some View {
        switch model.state.step {
        case OnboardingStep.insurance:
            documentStep(
                target: .insurance,
                slotLabelKey: "vehicle_onboard_insurance_slot",
                captionKey: "vehicle_onboard_insurance_caption"
            )

        case OnboardingStep.revenue:
            documentStep(
                target: .revenueLicence,
                slotLabelKey: "vehicle_onboard_revenue_slot",
                captionKey: "vehicle_onboard_revenue_caption"
            )

        case OnboardingStep.photos:
            photosStep

        // `details`, plus the arm a Kotlin enum forces on every Swift `switch` over one.
        default:
            detailsStep
        }
    }

    /// **SCR-DI-004 · Step 1/4** — vehicle type and registration number, and nothing else.
    ///
    /// The notice card is not decoration. It is the one place a Mode-C driver is told where a permit
    /// and a Mode A/B vehicle go, and the absence of those two fields is the AL-27 fence itself.
    private var detailsStep: some View {
        VStack(alignment: .leading, spacing: MageRideSpacing.sm) {
            GroupedList {
                GroupedRow(titleKey: "vehicle_onboard_type_label") {
                    vehicleTypeMenu
                }
                GroupedRow(titleKey: "vehicle_onboard_registration_label", showsSeparator: false) {
                    TextField(
                        "",
                        text: Binding(
                            get: { model.state.registrationNumber },
                            set: model.onRegistrationChanged
                        )
                    )
                    .mageFont(.body)
                    .foregroundStyle(MageRideColor.onSurface)
                    .multilineTextAlignment(.trailing)
                    .textInputAutocapitalization(.characters)
                    .autocorrectionDisabled()
                    .accessibilityLabel(Text(key: "vehicle_onboard_registration_label"))
                }
            }

            if model.state.isRegistrationTaken {
                FormErrorText(messageKey: "error_registration_exists")
            }

            NoticeCard(
                titleKey: "vehicle_onboard_mode_c_title",
                symbolName: "info.circle.fill",
                accent: MageRideVehicleColor.modeC
            ) {
                Text(key: "vehicle_onboard_mode_c_body")
                    .mageFont(.bodySmall)
                    .foregroundStyle(MageRideColor.onSurface)
            }
        }
    }

    /// The Mode-C subset (AL-09), as the wireframe's `Sedan ›` row.
    ///
    /// A `Menu` rather than the Android screen's dropdown text field, because `driver_ios.html` draws
    /// this as a grouped-list row with a chevron and D2' §C maps that onto `Form`. `bus` and `train`
    /// are absent by construction — the values are `RideVehicleType`s, and `POST /v1/vehicles`
    /// answers `403 mode-not-allowed` for either.
    private var vehicleTypeMenu: some View {
        Menu {
            ForEach(Self.modeCTypes, id: \.wire) { type in
                Button(action: { model.onVehicleTypeChanged(type) }) {
                    Text(key: type.labelKey)
                }
            }
        } label: {
            HStack(spacing: 2) {
                Text(model.state.vehicleType.map { $0.labelKey.localised } ?? Self.unset)
                    .mageFont(.body)
                    .foregroundStyle(MageRideColor.onSurface)
                Image(systemName: "chevron.right")
                    .font(.footnote)
                    .foregroundStyle(MageRideColor.outline)
            }
        }
        .accessibilityLabel(Text(key: "vehicle_onboard_type_label"))
    }

    /// **SCR-DI-004a / 004b · Steps 2 and 3** — one capture, one `Done ✓`, one extract card.
    ///
    /// The two steps differ only in what they capture and which fields come back, so they are one
    /// view. The extract card appears only after a save, because the save is what queues the Gemini
    /// Flash extraction — see ``VehicleOnboardingModel``.
    private func documentStep(
        target: DocumentCaptureTarget,
        slotLabelKey: String,
        captionKey: String
    ) -> some View {
        VStack(alignment: .leading, spacing: MageRideSpacing.sm) {
            CaptureTile(
                labelKey: captionKey,
                isCaptured: model.state.isCaptured(target),
                height: MageRideControl.capturePanel,
                onTap: { capture(target) }
            )

            GroupedList {
                GroupedRow(titleKey: slotLabelKey, showsSeparator: false) {
                    StatusPill.captured(model.state.isCaptured(target))
                }
            }

            extractionCard(isEditable: true)
        }
    }

    /// **SCR-DI-004c · Step 4/4** — front and back, number plate visible on both.
    ///
    /// The notice under the tiles is the step's whole rule: the plate read out of these photos is
    /// matched against the registration number typed on Step 1/4, and a mismatch is what makes the
    /// step Pending (BR-25.3, US-2.27).
    private var photosStep: some View {
        VStack(alignment: .leading, spacing: MageRideSpacing.sm) {
            HStack(spacing: MageRideSpacing.xs) {
                CaptureTile(
                    labelKey: "vehicle_onboard_photo_front",
                    isCaptured: model.state.photoFront != nil,
                    height: MageRideControl.capturePanel,
                    onTap: { capture(.vehicleFront) }
                )
                CaptureTile(
                    labelKey: "vehicle_onboard_photo_back",
                    isCaptured: model.state.photoBack != nil,
                    height: MageRideControl.capturePanel,
                    onTap: { capture(.vehicleBack) }
                )
            }

            GroupedList {
                GroupedRow(titleKey: "vehicle_onboard_photos_slot", showsSeparator: false) {
                    StatusPill.captured(model.state.photoFront != nil && model.state.photoBack != nil)
                }
            }

            NoticeCard(symbolName: "info.circle.fill", accent: MageRideVehicleColor.modeC) {
                Text("vehicle_onboard_plate_match".localisedFormat(model.state.registrationNumber))
                    .mageFont(.bodySmall)
                    .foregroundStyle(MageRideColor.onSurface)
            }

            // Not editable: Step 4/4's two rows are the plate check itself and are never the
            // driver's to retype — see ``VehicleFieldKeys/correctable``.
            extractionCard(isEditable: false)
        }
    }

    private func capture(_ target: DocumentCaptureTarget) {
        model.requestCapture(target)
        onCaptureRequested()
    }

    // MARK: - The extract card

    /// The wireframe's *"✦ AI-extracted (Gemini Flash 3.0)"* card, with the ✎ it draws.
    ///
    /// **Δ MCS-02 — the rows are editable.** `OnboardingCorrections` carries the driver's value and
    /// the document already on record is corrected in place, with no second photograph. Only the keys
    /// the step's document kind accepts get a ✎: `plate_text` and `reg_no_match` are **not** among
    /// them, because they are the fraud check Step 4/4 exists to perform and a driver who could
    /// retype the plate the camera read would be verifying their own vehicle.
    @ViewBuilder
    private func extractionCard(isEditable: Bool) -> some View {
        let fields = model.state.stepFields
        if !fields.isEmpty {
            NoticeCard(
                titleKey: "vehicle_onboard_extract_title",
                symbolName: "sparkles",
                accent: MageRideColor.success
            ) {
                ForEach(fields, id: \.key) { field in
                    extractedRow(field, isEditable: isEditable)
                }

                // The wireframe's `Confidence · 0.62 — doubtful` row. "Doubtful" is read off the
                // server's own `verifyStatus` rather than compared against a threshold here:
                // `Registry:OcrConfidenceThreshold` is admin-tunable and a second copy of it in the
                // app would eventually disagree with the verdict printed beside it.
                if let confidence = model.state.lowestConfidence {
                    ExtractedFieldRow(
                        labelKey: "vehicle_field_confidence",
                        value: fields.contains(where: \.needsOfficerReview)
                            ? "vehicle_confidence_doubtful".localisedFormat(formattedConfidence(confidence))
                            : formattedConfidence(confidence),
                        isFlagged: false
                    )
                }
            }
        }
    }

    @ViewBuilder
    private func extractedRow(_ field: ExtractedField, isEditable: Bool) -> some View {
        let canEdit = isEditable && VehicleFieldKeys.correctable.contains(field.key)

        ExtractedFieldRow(
            labelKey: field.labelKey,
            value: model.state.corrections[field.key] ?? field.displayValue,
            isFlagged: field.needsOfficerReview,
            onEdit: canEdit ? { model.toggleEdit(field.key) } : nil
        )

        if model.state.editingKey == field.key, canEdit {
            LabelledTextField(
                labelKey: field.labelKey,
                value: Binding(
                    get: { model.state.corrections[field.key] ?? field.value ?? "" },
                    set: { model.onCorrectionChanged(key: field.key, value: $0) }
                ),
                supportingKey: "vehicle_onboard_correction_supporting"
            )
        }
    }

    /// The wireframe's amber *"Any element doubtful or edited → this step's status is Pending"* card.
    private var pendingReviewCard: some View {
        NoticeCard(
            titleKey: "vehicle_onboard_pending_title",
            symbolName: "exclamationmark.triangle.fill",
            accent: MageRideColor.warning,
            fill: MageRideColor.warning.opacity(0.12)
        ) {
            HStack(alignment: .top, spacing: MageRideSpacing.xs) {
                Text(key: "vehicle_onboard_pending_body")
                    .mageFont(.bodySmall)
                    .foregroundStyle(MageRideColor.onSurface)
                Spacer(minLength: 0)
                AdminVerifyChip()
            }
        }
    }

    // MARK: -

    /// AL-09's eight ride-bookable types, in the wireframe's own order (the tier ladder, motorbike
    /// through mini truck) rather than the wire enum's.
    ///
    /// **The wireframe's caption under this row lists ten**, `bus` and `train` included, and that
    /// line is deliberately not rendered: those two are Mode A, `POST /v1/vehicles` answers
    /// `403 mode-not-allowed` for either, and offering one would be a control that can only fail. The
    /// Android screen offers the same eight. See the C087 handoff.
    private static let modeCTypes: [RideVehicleType] = [
        RideVehicleType.motorbike,
        RideVehicleType.threeWheeler,
        RideVehicleType.flex,
        RideVehicleType.sedan,
        RideVehicleType.miniVan,
        RideVehicleType.van,
        RideVehicleType.truck,
        RideVehicleType.miniTruck,
    ]

    /// What the type row shows before a type is chosen. An em dash rather than a string key, for
    /// `LocalizationTests`' own reason: a glyph is the same in all three scripts and three identical
    /// values in three strings files is exactly what that test (correctly) fails on.
    private static let unset = "—"
}
